plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -0,0 +1,274 @@
+			continue;
+		built_path = g_build_path ("/", basepath,
+					   "force_power", NULL);
+		if (g_file_test (built_path, G_FILE_TEST_IS_REGULAR)) {
+			data->force_path = g_steal_pointer (&built_path);
+			g_debug ("Detected force power support at %s",
+				 data->force_path);
+			break;
+		}
+	}
+	g_list_foreach (devices, (GFunc) g_object_unref, NULL);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_supported (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	return (data->force_path != NULL);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_set (FuPlugin *plugin, gboolean enable,
+				 GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	gint fd;
+	gint ret;
+
+	if (!fu_plugin_thunderbolt_power_supported (plugin)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "unable to set power to %d (missing kernel support)",
+			     enable);
+		return FALSE;
+	}
+	g_debug ("Setting force power to %d", enable);
+	fd = g_open (data->force_path, O_WRONLY);
+	if (fd == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "failed to open %s", data->force_path);
+		return FALSE;
+	}
+	ret = write (fd, enable ? "1" : "0", 1);
+	if (ret < 1) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "could not write to force_power': %s",
+			     g_strerror (errno));
+		g_close (fd, NULL);
+		return FALSE;
+	}
+	return g_close (fd, error);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_reset_cb (gpointer data)
+{
+	if (!fu_plugin_thunderbolt_power_set (FU_PLUGIN (data), FALSE, NULL))
+		g_warning ("failed to reset thunderbolt power");
+	return FALSE;
+}
+
+static gboolean
+udev_uevent_cb (GUdevClient *udev,
+		const gchar *action,
+		GUdevDevice *device,
+		gpointer     user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN(user_data);
+
+	if (action == NULL)
+		return TRUE;
+
+	g_debug ("uevent for %s: %s", g_udev_device_get_sysfs_path (device), action);
+
+	/* intel-wmi-thunderbolt has been loaded/unloaded */
+	if (g_str_equal (action, "change")) {
+		fu_plugin_thunderbolt_power_get_path (plugin);
+		fu_plugin_set_enabled (plugin, TRUE);
+		fu_plugin_recoldplug (plugin);
+	}
+
+	return TRUE;
+}
+
+/* virtual functions */
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+	const gchar *subsystems[] = { "wmi", NULL };
+
+	data->udev = g_udev_client_new (subsystems);
+	g_signal_connect (data->udev, "uevent",
+			  G_CALLBACK (udev_uevent_cb), plugin);
+	/* initially set to true, will wait for a device_register to reset */
+	data->needs_forcepower = TRUE;
+
+	/* determines whether to run device_registered */
+	fu_plugin_thunderbolt_power_get_path (plugin);
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_object_unref (data->udev);
+	if (data->force_path)
+		g_free (data->force_path);
```

Comment:
you don't need the `if` here, `g_free(NULL)` is explicitly okay.

Reply:
Fixed